src/cache/cache_control/cache_control.rs

```diff
@@ -0,0 +1,212 @@
+use crate::cache::CacheDirective;
+use crate::headers::{HeaderName, HeaderValue, Headers, ToHeaderValues, CACHE_CONTROL};
+
+use std::fmt::{self, Debug, Write};
+use std::iter::Iterator;
+use std::option;
+use std::slice;
+
+/// A Cache-Control header.
+///
+/// # Examples
+///
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::Response;
+/// use http_types::cache::{CacheControl, CacheDirective};
+/// let mut entries = CacheControl::new();
+/// entries.push(CacheDirective::Immutable);
+/// entries.push(CacheDirective::NoStore);
+///
+/// let mut res = Response::new(200);
+/// entries.apply(&mut res);
+///
+/// let entries = CacheControl::from_headers(res)?.unwrap();
+/// let mut entries = entries.iter();
+/// assert_eq!(entries.next().unwrap(), &CacheDirective::Immutable);
+/// assert_eq!(entries.next().unwrap(), &CacheDirective::NoStore);
+/// #
+/// # Ok(()) }
+/// ```
+pub struct CacheControl {
+    entries: Vec<CacheDirective>,
+}
+
+impl CacheControl {
+    /// Create a new instance of `ServerTiming`.
+    pub fn new() -> Self {
+        Self { entries: vec![] }
+    }
+
+    /// Create a new instance from headers.
+    pub fn from_headers(headers: impl AsRef<Headers>) -> crate::Result<Option<Self>> {
```

Comment:
I thought `from_headers()` was supposed to just be `Option<Self>`?

And this doesn't actually seem to make use of `Result`?

(Although I think that in many cases `Result<Option<Header>>` will probably make sense but is so far inconsistent...)

Reply:
`from_headers` needs to be able to both detect whether a header was found, and whether parsing the header was successful. These are different code paths, and probably should be handled separately.

However in #99 an argument is made to use `Option<Result<Self>>` instead, which may actually be better. We should probably adjust all existing `from_header` methods to this shape, and update this one too (the only existing ones are currently behind an unstable flag, so we're okay).

Reply:
On further thought I think that `Result<Option<T>>` should probably always be preferred, because it allows a seamless transition to `-> Option<T> throws Result` if that ever becomes a thing.

Reply:
The right signature would be `-> Option<T> throws Error`, but even then it'd be possible to write `-> crate::Result<T> throws NoneError`. I don't think the possibility of `throws` is something we should consider as a constraint for any design now. It's unclear if we'll ever have it, and even then the semantics are still undefined.